### PR TITLE
Refactoring of compressed filter pushdown

### PR DESCRIPTION
**Pull request by akuzm**
Bring qual_pushdown_mutator() and the individual pushdown functions to the same interface as expression_tree_mutator() (no NULL returns + using context.can_pushdown) to avoid accidental errors.

Reorganize the code a little to make adding new pushdown optimizations easier.

Fix a couple of bugs where we apply the original operator to the pushed-down clause that needs rechecking. This is not correct because the operator can turn false positives into false negatives.

Disable-check: force-changelog-file

### CODE REVIEW

**Review by svenklemm — APPROVED** [review]

File: tsl/src/nodes/decompress_chunk/qual_pushdown.c [hunk 1]
```diff
@@ -209,40 +191,48 @@ pushdown_op_to_segment_meta_min_max(QualPushdownContext *context, List *expr_arg
...
 
 	/* May be able to allow non-strict operations as well.
 	 * Next steps: Think through edge cases, either allow and write tests or figure out why we must
 	 * block strict operations
 	 */
 	if (!OidIsValid(op_oid) || !op_strict(op_oid))
-		return NULL;
+	{
+		context->can_pushdown = false;
+		return orig_opexpr;
```
Comment (mkindahl) [inline tsl/src/nodes/decompress_chunk/qual_pushdown.c:220]: We are lacking tests for this case, but it seems like the plan is to write tests in a separate PR.

File: tsl/src/nodes/decompress_chunk/qual_pushdown.c [hunk 2]
```diff
@@ -343,18 +343,21 @@ expr_fetch_bloom1_metadata(QualPushdownContext *context, Expr *expr, AttrNumber
...
 	Expr *original_leftop;
 	Expr *original_rightop;
 	TypeCacheEntry *tce;
 	int strategy;
 
-	if (list_length(expr_args) != 2)
-		return NULL;
-
+	List *expr_args = orig_opexpr->args;
+	Assert(list_length(expr_args) == 2);
```
Comment (mkindahl) [inline tsl/src/nodes/decompress_chunk/qual_pushdown.c:387]: What is the operator here? Do you know that it is always guaranteed to have 2 arguments?

If not, we will have weird errors in production builds since it will continue execution under the assumption that there are two arguments. If fewer, we might have crashes. If more, we could have strange results.

Might be better to use 'Ensure` here to generate errors if something changes or we don't have good test coverage.

File: tsl/src/nodes/decompress_chunk/qual_pushdown.c [hunk 3]
```diff
@@ -547,11 +532,122 @@ modify_expression(Node *node, QualPushdownContext *context)
...
+			}
+
+			if (opexpr->opresulttype != BOOLOID)
+			{
+				/*
+				 * The following pushdown options only support operators that
+				 * return bool.
+				 */
+				context->can_pushdown = false;
+				return orig_node;
```
Comment (mkindahl) [inline tsl/src/nodes/decompress_chunk/qual_pushdown.c:597]: We don't seem to have coverage for operators returning bool?!

**Review by mkindahl — APPROVED** [review]
I'm a little worried that we have such bad coverage in the code, but this is not something you introduced so not going to ask you to fix this in this PR.

File: tsl/src/nodes/decompress_chunk/qual_pushdown.c (hunk 1 above)
Comment (akuzm) [reply tsl/src/nodes/decompress_chunk/qual_pushdown.c:220]: This is probably impossible to cover, this requires a custom broken type that is missing some btree strategies, but we won't be able to create min/max columns then. This is just to be on the safe side in case there's some catalog corruption.

File: tsl/src/nodes/decompress_chunk/qual_pushdown.c (hunk 2 above)
Comment (akuzm) [reply tsl/src/nodes/decompress_chunk/qual_pushdown.c:387]: This is checked in the caller properly already, this is just a precondition check for this function.

File: tsl/src/nodes/decompress_chunk/qual_pushdown.c (hunk 1 above)
Comment (mkindahl) [reply tsl/src/nodes/decompress_chunk/qual_pushdown.c:220]: Yes, that makes sense. Just worried that there is some corner-case that we haven't covered. Not much to do in this PR though.

File: tsl/src/nodes/decompress_chunk/qual_pushdown.c (hunk 3 above)
Comment (akuzm) [reply tsl/src/nodes/decompress_chunk/qual_pushdown.c:597]: The other way around. We can only have the operators that return bool in the where clause, so other operators are not checked.
